Best practices for developing recognizers doc (#503)
microsoft/presidio: docs/analyzer/developing_recognizers.md

# Recognizers Development - Best Practices and Considerations

Recognizers are the main building blocks in Presidio. Each recognizer is in charge of detecting one or more entities in one or more languages.
Recognizers define the logic for detection, as well as the confidence a prediction receives and a list of words to be used when context is leveraged.

## Implementation Considerations

### Accuracy

Each recognizer, regardless of its complexity, could have false positives and false negatives. When adding new recognizers, we try to balance the effect of each recognizer on the entire system. A recognizer with many false positives would affect the system's usability, while a recognizer with many false negatives might require more work before it can be integrated. For reproducibility purposes, it is be best to note how the recognizer's accuracy was tested, and on which datasets.

> For tools and documentation on evaluating and analyzing recognizers, refer to the [presidio-research Github repository](https://www.github.com/presidio-research).

Reviewer: This - and the two following links give me a 404 - is this a private repo perhaps?

Author: Thanks! fixed the github link. The others are causing issues because the branch name has a / in it. I'll verify this with a different branch.

### Pattern Based

Pattern based recognizers use regular expressions to identify entities in text.
See [this documentation](/docs/custom_fields.md#via-code) on adding a new recognizer via code.
The [`PatternRecognizer`](/presidio-analyzer/presidio_analyzer/pattern_recognizer.py) class should be extended.
See some examples here:

> Examples of pattern based recognizers are the [`CreditCardRecognizer`](../../presidio-analyzer/presidio_analyzer/predefined_recognizers/credit_card_recognizer.py) and [`EmailRecognizer`](../../presidio-analyzer/presidio_analyzer/predefined_recognizers/email_recognizer.py).

### Machine Learning (ML) Based or Rule-Based

Many PII entities are undetectable using naive approaches like deny-lists or regular expressions.
In these cases, we would wish to utilize a Machine Learning model capable of identifying entities in free text, or a rule-based recognizer. There are four options for adding ML and rule based recognizers:

#### Utilize SpaCy or Stanza

> Presidio currently uses [spaCy](https://spacy.io/) as a framework for text analysis and Named Entity Recognition (NER), and [stanza](https://stanfordnlp.github.io/stanza/) as an alternative. As to not introduce new tools for no reason, it is recommended to first try to use `spaCy` or `stanza` over other tools.

Reviewer: ```suggestion
Presidio currently uses [spaCy](https://spacy.io/) as a framework for text analysis and Named Entity Recognition (NER), and [stanza](https://stanfordnlp.github.io/stanza/) as an alternative. To avoid introducing new tools, it is recommended to first try to use `spaCy` or `stanza` over other tools if possible.
```

The Presidio `EntityRecognizer` API allows you to use `spaCy`/`stanza` extracted features like lemmas, part of speech, dependencies and more to create your logic. When integrating such logic into Presidio, a class inheriting from the [`EntityRecognizer`](../../presidio-analyzer/presidio_analyzer/entity_recognizer.py) should be created.

#### Deep Learning Based Methods

Deep learning methods offer excellent detection rates for NER.
They are however more complex to train, deploy and tend to be slower than traditional approaches.
When creating a DL based method for PII detection, there are two main alternatives for integrating it with Presidio:

1. Create an external endpoint (either local or remote) which is isolated from the `presidio-analyzer` process. On the `presidio-analyzer` side, one would extend the [`RemoteRecognizer`](../../presidio-analyzer/presidio_analyzer/remote_recognizer.py) class and implement the network interface between `presidio-analyzer` and the endpoint of the model's container.
2. Integrate the model as an additional [`EntityRecognizer`](../../presidio-analyzer/presidio_analyzer/entity_recognizer.py) within the `presidio-analyzer` flow.

> > Considerations for selecting one option are, among others: Ease of integration, runtime considerations (For example if the new model requires a GPU) and the 3rd party dependencies of the new model vs. the existing `presidio-analyzer` package.

Reviewer: ```suggestion
> Considerations for selecting one option over another are: Ease of integration, runtime considerations (For example if the new model requires a GPU) and the 3rd party dependencies of the new model vs. the existing `presidio-analyzer` package.
```
